File: w06/d02/INSTRUCTORS/callbacks/callbacks.js
var runTrue = function () { console.log("Success!"); };
var runFalse = function () { console.log("Failure!"); };

function doIt(onSuccess, onFailure) {
  var success = Math.floor(Math.random() * 10);
  if (success) {
    onSuccess();
  } else {
    onFailure();
  }
}

doIt(runTrue, runFalse);


// MOAR COMPLICATED!

var onSuccess = "Success!";
var onFailure = "Failure!";
var callback = function callback(msg, time) {
  console.log(msg + " at " + time);
};

function doIt(onSuccess, onFailure, callback) {
  function performOperation(workToDoAfterDone) {
    setTimeout(workToDoAfterDone, 1000);
  }
  function theWorkIamDoing() {
    var success = Math.floor(Math.random() * 10);
    // A wild operator appears! Google "ternary operator"
    var message = success ? onSuccess : onFailure;
    callback(message, Date.now());
  }
  performOperation(theWorkIamDoing);
  console.log("Please be patient, this may take a few seconds");
}

doIt(onSuccess, onFailure, callback);

File: w06/d02/INSTRUCTORS/callbacks/callbacks.rb
run_true = Proc.new { puts "Success! #{Time.now - 1_500}" }
run_false = Proc.new { puts "Failure! #{puts Time.now + 1_500_000}" }

def do_it on_success, on_failure
  success = rand(10) < 5
  if success
    on_success.call
  else
    on_failure.call
  end
end

do_it(run_true, run_false)


### MOAR COMPLICATED!

require 'thread'

on_success = "Success!"
on_failure = "Failure!"
callback = Proc.new { |msg, time| puts "#{msg} at #{time}" }
def do_it on_success, on_failure, callback
  def perform_operation
    sleep rand*5
  end
  Thread.new do
    success = rand(10) < 5
    # A wild operator appears! Google "ternary operator"
    message = success ? on_success : on_failure
    perform_operation
    callback.call(message, Time.now)
  end
  puts "Be patient, this may take a few seconds."
end

do_it(on_success, on_failure, callback)

# Try to describe what is going on!
